### PR TITLE
Fix PFTrackAlgoTools memory corruption

### PR DESCRIPTION
**Pull request by kpedro88**
A memory corruption issue in `PFTrackAlgoTools` was identified in #16493. Two functions in this namespace referenced indices [5] and [6] from an input vector of cut values. However, [5] only exists sometimes and [6] never exists:
https://cmssdt.cern.ch/lxr/search?_filestring=&_string=NHitCuts_byTrackAlgo
https://cmssdt.cern.ch/lxr/search?_filestring=&_string=DPtOverPtCuts_byTrackAlgo

@bachtis suggested just using `back()` for all cases that reference [5] or [6]. This indeed fixes the memory corruption issue (checked with valgrind).

This PR will be backported to 81X.

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
This PR is making extensive changes in Phase 2 workflows. Differences show up in the Jenkins test resutls. Also, I tested workflow 20821.0_TenMuE_0_200+TenMuE_0_200_pythia8_2023D3 with 200 events against baseline CMSSW_9_0_X_2016-11-28-1100. There are many changes, and I don't understand how a seemingly innocuous bug fix could cause such changes. A sample is shown below. (Red is the PR, black is the baseline.)

all_testpr16797vsorig_tenmue2023d3wf20821p0c_log10recopfcandidates_particleflow__reco_obj_p53
all_testpr16797vsorig_tenmue2023d3wf20821p0c_recopfcandidates_particleflow__reco_obj_eta
angnijiets
pujets

momni
mueff-1
mueta-1

**Comment by kpedro88** [comment]
@cvuosalo I was surprised by this as well, so I checked out a clean IB and added cout statements wherever the affected vector entries were used. Here is a representative sample from 1 event in 20821:
```
dPtCut(): cuts[5] = 6.94908e-310
nHitCut(): cuts[5] = 32747
dPtCut(): cuts[5] = 6.94908e-310
nHitCut(): cuts[5] = 32747
dPtCut(): cuts[5] = 6.94908e-310
nHitCut(): cuts[5] = 32747
```
So the out-of-bounds values were indeed being used quite frequently.

**Comment by slava77** [comment]
On 11/29/16 9:20 PM, Kevin Pedro wrote:
> @cvuosalo <https://github.com/cvuosalo> I was surprised by this as well,
> so I checked out a clean IB and added cout statements wherever the
> affected vector entries were used. Here is a representative sample from
> 1 event in 20821:
> 
> |dPtCut(): cuts[5] = 6.94908e-310 nHitCut(): cuts[5] = 32747 dPtCut():
> cuts[5] = 6.94908e-310 nHitCut(): cuts[5] = 32747 dPtCut(): cuts[5] =
> 6.94908e-310 nHitCut(): cuts[5] = 32747 |
> 
> So the out-of-bounds values were indeed being used quite frequently.


These cuts apparently work for 2016 setup.
Does it mean that the phase-2 code is not configured correctly?
Just using the last filled value may be OK from avoiding bad memory
reads POV,
but is it the right value?

**Comment by slava77** [comment]
@kpedro88 
please add module/instance names for the cuts[5] .. and maybe cuts[6] in the printouts (unless there is just a single module for this).
This should point to the corresponding configs.
If code expects that a specific iteration is cut in a specific index, then we should have appropriate input configs passed in.

**Comment by lgray** [comment]
Doesn't phase 2 have new tracking iterations, perhaps there's some alignment of cuts that needs to be done? It's incredible that anything was working before...

The complete removal of muons in the endcap I really don't understand...

**Comment by makortel** [comment]
> Doesn't phase 2 have new tracking iterations

Phase2 iteration names are a subset of those in phase1, and those were added in this file already before they were taken into use in phase1 tracking. The (simple) idea back then was to put the new quadruplet iterations to the same category as the phase0 triplet iterations. Of course that does not mean that the cut values would be optimal.

Anyway, it seems that e.g. `PFTrackAlgoTools::goodPtResolution()` (which calls `nHitCut()`) is called from `GeneralTracksImporter(WithVeto)` and `HGCalTrackCollectionProducer`. Looking their configurations
[RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py](https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py#L41)
[RecoParticleFlow/PFTracking/python/hgcalTrackCollection_cfi.py](https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/RecoParticleFlow/PFTracking/python/hgcalTrackCollection_cfi.py)
the HGCal one has indeed one element less (same holds for `DPtOverPtCuts_byTrackAlgo`). Is this difference really intended (or could it be e.g. a copy-paste mistake)?

For a PF-outsider the fact that the cut vector has "sometimes 5 and sometimes 6 elements" sounds a bit strange (and error prone).

**Comment by lgray** [comment]
@makortel Thanks for taking a look! Indeed the ```HGCalTrackCollectionProducer``` came from SLHC releases directly which (probably?) had a different scheme. We should likely simply update the config for that producer. This will likely fix the muons for HGCal.

**Comment by lgray** [comment]
@kpedro88 When you get this could you copy/paste the config from PFBlockAlgo for the track selection into the ```HGCalTrackCollectionProducer``` config?

**Comment by makortel** [comment]
If the config copy is the correct fix and `PFTrackAlgoTools` can be assumed to be correct, how about adding a check (possibly abstracked to `PFTrackAlgoTools` and throwing an exception) everywhere taking these cuts from configuration to ensure that the cut vectors have 6 elements?

**Comment by lgray** [comment]
The HLT configs have fewer (different track iterations).

On Wed, Nov 30, 2016 at 9:59 AM, Matti Kortelainen <notifications@github.com

**Comment by makortel** [comment]
I see `hltIterativeTracking` flag, which must be a configuration variable too. So the check could require 3 elements instead of 7 if that flag is set.

**Comment by bachtis** [comment]
Hi Just to confirm what @makortel is saying. We added the changes in 8XY release so when HGC got forwarded it should be updated to the new track tools. This was done to allow some cleaning of noisy tracks in 13 TeV data

**Comment by slava77** [comment]
On 11/30/16 12:31 AM, Matti Kortelainen wrote:
> For a PF-outsider the fact that the cut vector has "sometimes 5 and
> sometimes 6 elements" sounds a bit strange (and error prone).
> 

looks like a good place to change [5] to .at(5)

**Comment by kpedro88** [comment]
@makortel well, part of the problem is that the cut vectors *never* have 7 elements, so the use of `cuts[6]` never really made sense.

I can update the HGCal config and change `.back()` to `.at(5)` as a basic safety check, but maybe this still needs a bit more thought from PF experts.

**Comment by kpedro88** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc [hunk 1]
```diff
@@ -36,7 +36,7 @@ namespace PFTrackAlgoTools {
     case reco::TrackBase::hltIterX:
       return  cuts[0];
     default:
-      return hltIterativeTracking ? cuts[6]:cuts[0];
+      return hltIterativeTracking ? cuts.at(5):cuts[0];
```
Comment (slava77) [inline RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc]: @bachtis @fwyzard is it correct to set this to 5 for hlt highest iteration while we use 0 for the offline tracking highest  (or not implemented yet) iteration?

Maybe an exception or something that means no cut should be placed instead.

**Comment by kpedro88** [comment]
Comparisons look healthier now, though there are still some minor fluctuations in 2023 PF plots.

**Comment by cvuosalo** [comment]
New test of workflow 20821.0 is in progress...

File: RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc (hunk 1 above)
Comment (fwyzard) [inline RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc]: I have no idea what this is about...

Comment (fwyzard) [inline RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc]: I find this way of passing and applying cuts (in a vector, with a magic way of picking the index) to defeat the whole purpose of having variable names and a configuration language.

Comment (slava77) [inline RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc]: @mtosi please check this

**Comment by cvuosalo** [comment]
I repeated my test of 20821.0 (https://github.com/cms-sw/cmssw/pull/16797#issuecomment-263737458) and still get the same results. I don't see any improvement.

**Comment by kpedro88** [comment]
I guess we need feedback from PF experts, then. This PR just avoids using uninitialized/out-of-bounds values, and apparently updating the HGC cuts had no effect. If previous expectations of performance were based on nonsense cuts, then expectations need to be revised...

**Comment by cvuosalo** [comment]
@kpedro88: If this PR needs to go into 900pre2, we should resolve the issues now. After revision, I'll have to run my tests again, so some time for additional review is still required before this PR can be merged.

**Comment by kpedro88** [comment]
I agree, but I'm not sure how to resolve them. We need input from @bachtis or other experts.

**Comment by davidlange6** [comment]
adding also crosspog - @arizzi @gpetruc 

however - in the context of phase2 physics changes - @kpedro88 - why the numerical changes to the values in DPtOverPtCuts_byTrackAlgo in this PR? Can we separate the two issues?

**Comment by kpedro88** [comment]
The changes in values for HGC were recommended by @lgray and @bachtis earlier in the comments on this PR

**Comment by davidlange6** [comment]
ok - now I see - but those are presumably the source of the physics changes, right? Using at should ensure we aren't overflowing. Only the 6->5 part doesn't make sense (but never did)..

**Comment by kpedro88** [comment]
No, we made those changes after Carl made the plots, in hopes of improving things.

**Comment by davidlange6** [comment]
ok, I didn’t notice that from the history. I’m confused how this PR can change something that is apparently desirable without those changes.

**Comment by kpedro88** [comment]
To me it looks like the use of out-of-bounds values: https://github.com/cms-sw/cmssw/pull/16797#issuecomment-263785852

**Comment by davidlange6** [comment]
Sure - but code accessing those should not have resulted in a useful output before..

**Comment by cvuosalo** [comment]
@kpedro88: The window is rapidly closing for this PR to get into 900pre2. We need major progress by Wednesday for there to be much of a chance of inclusion.

**Comment by kpedro88** [comment]
@cvuosalo I'm not sure what you want me to do. I'm not a PF expert and those who are don't seem to be responding. I don't know what should be the right answer for this code from a physics perspective; I only know that referencing nonexistent vector indices is a bad idea.

**Comment by slava77** [comment]
On 12/6/16 5:27 PM, Kevin Pedro wrote:
> @cvuosalo <https://github.com/cvuosalo> I'm not sure what you want me to
> do. I'm not a PF expert and those who are don't seem to be responding. I
> don't know what should be the right answer for this code from a physics
> perspective; I only know that referencing nonexistent vector indices is
> a bad idea.

Can you change the parameters to match the old performance,
just to see that initialized values compatible with your printouts give
the same results?

**Comment by cvuosalo** [comment]
I am inclined to let this PR go in. Without the fix, we have random behavior when performing out-of-bounds array accesses. The fix may be revealing previously hidden bugs, but, at least, we are removing random behavior.

**Comment by slava77** [comment]
better understand the problem.
We have until the end of the week.

**Comment by kpedro88** [comment]
Here is a little more info from a clean IB:
```
dPtCut(): cuts[5] = 6.94517e-310
(algo = 13)
(cuts = -1,-1,-1,1,1,)
nHitCut(): cuts[5] = 32729
(algo = 13)
(cuts = 3,3,3,3,3,)
```
Algo 13 corresponds to `reco::TrackBase::muonSeededStepInOut`. It appears to be the only algo which tries to use nonexistent values. The actual values of the cut vectors do appear to come from the HGCal config, so I tried setting cuts[5] = 0 or 32700 (for dPt and nHit, respectively) in that config. Through a lot of trial and error, I've determined that it's only changing the last nHit cut value from ~32700 -> 3 that causes the loss of muons in the endcap.

Therefore, I've restored this (to me nonsensical) value in the latest commit. If we confirm this undoes all the apparent regressions, then maybe there can be a followup PR to tune this value to something sensible.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
On 12/7/16 3:27 PM, Kevin Pedro wrote:
> Here is a little more info from a clean IB:
> 
> |dPtCut(): cuts[5] = 6.94517e-310 (algo = 13) (cuts = -1,-1,-1,1,1,)
> nHitCut(): cuts[5] = 32729 (algo = 13) (cuts = 3,3,3,3,3,) |
> 
> Algo 13 corresponds to |reco::TrackBase::muonSeededStepInOut|. It
> appears to be the only algo which tries to use nonexistent values. The
> actual values of the cut vectors do appear to come from the HGCal
> config, so I tried setting cuts[5] = 0 or 32700 (for dPt and nHit,
> respectively) in that config. Through a lot of trial and error, I've
> determined that it's only changing the last nHit cut value from ~32700
> -> 3 that causes the loss of muons in the endcap.
> 
> Therefore, I've restored this (to me nonsensical) value in the latest
> commit. If we confirm this undoes all the apparent regressions, then
> maybe there can be a followup PR to tune this value to something sensible.
> 

Makes sense.
Maybe you can also create an issue on github so that it stays on the to
do list.

**Comment by cvuosalo** [comment]
A test with the new cut is in progress...

**Comment by cvuosalo** [comment]
+1

For #16797 86bf6126adcbafe46f242a5eb0acf041c2faa72f

Fixing out-of-bounds array access in PFTrackAlgoTools. This fix takes care of the original bugs but will require future tuning of the muon cut.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_9_0_X_2016-12-07-2300 show no significant differences, as might be expected. A  test of workflow 20821.0_TenMuE_0_200+TenMuE_0_200_pythia8_2023D3 using the latest version of the PR with 200 events against baseline CMSSW_9_0_X_2016-11-28-1100 shows only two, very tiny, insignificant differences in plots.

**Comment by slava77** [comment]
On 12/8/16 9:34 AM, Carl Vuosalo wrote:
> +1
> 
> For #16797 <https://github.com/cms-sw/cmssw/pull/16797> 86bf612
> <https://github.com/cms-sw/cmssw/commit/86bf6126adcbafe46f242a5eb0acf041c2faa72f>
> 
> Fixing out-of-bounds array access in PFTrackAlgoTools. This fix takes
> care of the original bugs but will require future tuning of the muon cut.
> 
> The code changes are satisfactory, and Jenkins tests against baseline
> CMSSW_9_0_X_2016-12-07-2300 show no significant differences, as might be
> expected. A test of workflow
> 20821.0_TenMuE_0_200+TenMuE_0_200_pythia8_2023D3 using the latest
> version of the PR with 200 events against baseline
> CMSSW_9_0_X_2016-11-28-1100 shows only two, very tiny, insignificant
> differences in plots.

Do we have an issue created on github to fix the cut behavior?

**Comment by kpedro88** [comment]
@slava77 I just submitted #16914

**Comment by slava77** [comment]
On 12/8/16 11:54 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> I just submitted #16914
> <https://github.com/cms-sw/cmssw/issues/16914>

Thank you.

**Comment by davidlange6** [comment]
+1

**Pull request merged**